ui/store/action-queue/index.test.js

```diff
@@ -0,0 +1,275 @@
+      const trace = {};
+      const background = {
+        connectionStream: {
+          readable: false,
+        },
+        first: (cb) => {
+          setTimeout(() => {
+            trace.firstDone = Date.now();
+            cb(null, 'first');
+          }, 5);
+        },
+        second: (cb) => {
+          trace.secondStarted = Date.now();
+          setTimeout(() => cb(null, 'second'), 10);
+        },
+      };
+      _setBackgroundConnection(background);
+      const scheduled = Promise.all([
+        submitRequestToBackground('first'),
+        submitRequestToBackground('second'),
+      ]);
+      background.connectionStream.readable = true;
+      _setBackgroundConnection(background);
+      await scheduled;
+      expect(trace.firstDone).toBeLessThanOrEqual(trace.secondStarted);
+    });
+
+    it('ensures actions in queue will not repeat once finished', async () => {
+      const trace = { calls: 0 };
+      const background = {
+        connectionStream: {
+          readable: false,
+        },
+        first: (cb) => {
+          trace.calls += 1;
+          setTimeout(() => {
+            trace.firstDone = Date.now();
+            cb(null, 'first');
+          }, 5);
+        },
+        second: (cb) => {
+          trace.calls += 1;
+          trace.secondStarted = Date.now();
+          setTimeout(() => cb(null, 'second'), 10);
+        },
+      };
+      _setBackgroundConnection(background);
+      const scheduled = Promise.all([
+        submitRequestToBackground('first'),
+        submitRequestToBackground('second'),
+      ]);
+      background.connectionStream.readable = true;
+      _setBackgroundConnection(background);
+      await scheduled;
+      _setBackgroundConnection(background); // once all actions finished, this triggers draining the queue again
+      expect(trace.firstDone).toBeLessThanOrEqual(trace.secondStarted);
+      expect(trace.calls).toStrictEqual(2);
+    });
+
+    it('stops processng the queue if connection is lost', async () => {
+      const trace = {};
+      const background = {
+        connectionStream: {
+          readable: false,
+        },
+        first: (cb) => {
+          setTimeout(() => {
+            trace.firstDone = true;
+            background.connectionStream.readable = false;
+            cb(Error('lost connection'));
+          }, 5);
+        },
+        second: sinon.stub().yields(),
+      };
+      _setBackgroundConnection(background);
+      const scheduled = Promise.race([
+        submitRequestToBackground('first').catch(() => ({})),
+        submitRequestToBackground('second'),
+      ]);
+      background.connectionStream.readable = true;
+      _setBackgroundConnection(background);
+      await scheduled;
+      await Promise.resolve('one more tick'); // One asynchronous tick to avoid depending on implementation details
+      expect(trace.firstDone).toStrictEqual(true);
+      expect(background.second.called).toStrictEqual(false);
+    });
+  });
```

Comment:
While testing I found that usually error is not thrown if connection is lost, just the action kind of hangs and there is no reply.

Reply:
I'll try to reproduce that too in a separate test